app/models/mixins/container_resource_mixin.rb

```diff
@@ -0,0 +1,141 @@
+    elsif error_code != 200
+      raise MiqException::MiqProvisionError, "Unexpected Exception while getting object with resource name #{name}, method name #{method_name}, HTTP response code #{error_code}, and message #{msg}."
+    end
+    response
+  end
+
+  def tidy_provider_definition
+    resource = provider_definition
+    resource[:metadata].delete(:selfLink)
+    resource[:metadata].delete(:uid)
+    resource[:metadata].delete(:resourceVersion)
+    resource[:metadata].delete(:creationTimestamp)
+    resource
+  end
+
+  private
+
+  # Retrieves the resource from the provider and returns the hash
+  def provider_definition(api_version = 'v1')
+    method_name = "get_#{kind_in_provider.underscore}"
+    response, error_code, msg = send_method(method_name, [name, namespace], api_version)
+    if error_code == 404
+      return nil
+    elsif error_code != 200
+      raise MiqException::MiqProvisionError, "Unexpected Exception while getting object with resource name #{name}, namespace #{namespace}, method name #{method_name}, HTTP response code #{error_code}, and message #{msg}."
+    end
+    response
+  end
+
+  MIQ_CLASS_MAPPING = {
+    "ContainerBuild"        => "BuildConfig",
+    "ContainerBuildPod"     => "Build",
+    "ContainerGroup"        => "Pod",
+    "ContainerLimit"        => "LimitRange",
+    "ContainerQuota"        => "ResourceQuota",
+    "ContainerReplicator"   => "ReplicationController",
+    "ContainerRoute"        => "Route",
+    "PersistentVolumeClaim" => "PersistentVolumeClaim",
+    "ContainerImage"        => "Image",
+    "ContainerService"      => "Service",
+  }.freeze
+
+  # Updates the resource in provider.  This is a wholesale replace.  The
+  # metadata/name and metadata/namespace are set to the name and namesapce of
+  # this resource.
+  def update_in_provider(resource, api_version = 'v1')
+    method_name = "update_#{kind_in_provider.underscore}"
+    resource.deep_symbolize_keys!
+    if resource[:kind] != kind_in_provider
+      raise MiqException::MiqProvisionError, "Unexpected Exception while updating object with resource name #{name} in namespace namespace #{namespace}.  The update kind #{resource[:kind]} doesn't match the existing resource kind #{kind_in_provider}."
+    end
+    params = [resource]
+    response, error_code, msg = send_method(method_name, params, api_version)
+    if error_code == 404
+      return nil
+    elsif error_code != 200
+      raise MiqException::MiqProvisionError, "Unexpected Exception while updating object with resource name #{name}, namespace #{namespace}, method name #{method_name}, HTTP response code #{error_code}, and message #{msg}."
+    end
+    response
+  end
+
+  # The send_method method actually invokes the Kubeclient call. The connect_client
+  # method obtains a client from the ext_management_system that is either configured
+  # for the OpenShift of Kubernetes API so that saves a step.
+  def send_method(method_name, params, api_version = 'v1')
```

Comment:
For example: why do we need to parse json here. It should be already taken care of, why do we have to catch the KubeException (just catch it where it's needed instead of the `if error_code`).